src/event.rs

```diff
@@ -627,6 +634,28 @@ pub struct KeyboardInput {
     pub modifiers: ModifiersState,
 }
 
+/// Describes an event from input method.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum Ime {
+    /// Notifies when the IME was enabled.
+    Enabled,
+
+    /// Notifies when a new composing text should be set at the cursor position.
+    ///
+    /// The value represents a pair of the preedit string and the cursor begin position and end
+    /// position. When both indices are `None`, the cursor should be hidden.
+    ///
+    /// The cursor position is byte-wise indexed.
+    Preedit(String, Option<usize>, Option<usize>),
```

Comment:
I think this should be `Preedit(String, Option<(usize, usize)>)`, since there is never a case (as far as I can see) where you know the beginning or the end, but not both.

(Even Wayland's [`Event::PreeditString`](https://docs.rs/wayland-protocols/0.29.4/wayland_protocols/unstable/text_input/v3/client/zwp_text_input_v3/enum.Event.html#variant.PreeditString) seem to indicate that both values are usually positive, except in the hidden case where they're both negative, though I'm not certain here at all!)

Reply:
Yeah, I'll change that.

Reply:
Done.